File: csharp/src/Vista.SDK/LocationBuilder.cs
using System.Runtime.CompilerServices;
using System.Text;
using Vista.SDK;

public sealed record LocationBuilder
{
    public int? Number { get; private init; }
    public char? Side { get; private init; }
    public char? Vertical { get; private init; }
    public char? Transverse { get; private init; }
    public char? Longitudinal { get; private init; }

    public VisVersion VisVersion { get; }

    private IReadOnlyDictionary<char, LocationGroup> _reversedGroups;

    private LocationBuilder(VisVersion visVersion, IReadOnlyDictionary<char, LocationGroup> reversedGroups)
    {
        VisVersion = visVersion;

        _reversedGroups = reversedGroups;
    }

    public static LocationBuilder Create(Locations locations) =>
        new LocationBuilder(locations.VisVersion, locations._reversedGroups);

    public LocationBuilder WithLocation(Location value)
    {
        var builder = this with { };

        var span = value.ToString().AsSpan();

        int? n = null;

        for (var i = 0; i < span.Length; i++)
        {
            ref readonly var ch = ref span[i];
            if (char.IsDigit(ch))
            {
                if (n is null)
                    n = ch - '0';
                else
                {
                    if (!Locations.TryParseInt(span, 0, i + 1, out var num))
                        throw new ArgumentException("Should include a valid number");
                    n = num;
                }
                continue;
            }

            builder = builder.WithValue(ch);
        }

        if (n is not null)
            builder = builder.WithNumber(n.Value);

        return builder;
    }

    public LocationBuilder WithNumber(int number) => WithValueInternal(LocationGroup.Number, number);

    public LocationBuilder WithSide(char side) => WithValueInternal(LocationGroup.Side, side);

    public LocationBuilder WithVertical(char vertical) => WithValueInternal(LocationGroup.Vertical, vertical);

    public LocationBuilder WithTransverse(char transverse) => WithValueInternal(LocationGroup.Transverse, transverse);

    public LocationBuilder WithLongitudinal(char longitudinal) =>
        WithValueInternal(LocationGroup.Longitudinal, longitudinal);

    public LocationBuilder WithValue(int value)
    {
        return WithValueInternal(LocationGroup.Number, value);
    }

    public LocationBuilder WithValue(char value)
    {
        if (!_reversedGroups.TryGetValue(value, out var key))
            throw new ValidationException($"The value {value} is an invalid Locations value");
        return WithValueInternal(key, value);
    }

    private LocationBuilder WithValueInternal<T>(LocationGroup group, T value)
        where T : struct
    {
        if (group == LocationGroup.Number)
        {
            if (typeof(T) != typeof(int))
                throw new ValidationException("Value should be number");
            ref var n = ref Unsafe.As<T, int>(ref value);
            if (n < 1)
                throw new ValidationException("Value should be greater than 0");
            return this with { Number = n };
        }

        if (typeof(T) != typeof(char))
            throw new ValidationException("Value should be a character");
        ref var val = ref Unsafe.As<T, char>(ref value);

        if (!_reversedGroups.TryGetValue(val, out var key) || key != group)
            throw new ValidationException(
                $"The value {value} is an invalid {Enum.GetName(typeof(LocationGroup), group)} value"
            );

        return group switch
        {
            LocationGroup.Side => this with { Side = val },
            LocationGroup.Vertical => this with { Vertical = val },
            LocationGroup.Transverse => this with { Transverse = val },
            LocationGroup.Longitudinal => this with { Longitudinal = val },
            _ => throw new Exception($"Unsupported code: {group}"),
        };
    }

    public LocationBuilder WithoutValue(LocationGroup group)
    {
        return group switch
        {
            LocationGroup.Number => this with { Number = null },
            LocationGroup.Side => this with { Side = null },
            LocationGroup.Vertical => this with { Vertical = null },
            LocationGroup.Transverse => this with { Transverse = null },
            LocationGroup.Longitudinal => this with { Longitudinal = null },
            _ => throw new Exception($"Unsupported code: {group}"),
        };
    }

    public LocationBuilder WithoutNumber() => WithoutValue(LocationGroup.Number);

    public LocationBuilder WithoutSide() => WithoutValue(LocationGroup.Side);

    public LocationBuilder WithoutVertical() => WithoutValue(LocationGroup.Vertical);

    public LocationBuilder WithoutTransverse() => WithoutValue(LocationGroup.Transverse);

    public LocationBuilder WithoutLongitudinal() => WithoutValue(LocationGroup.Longitudinal);

    public Location Build()
    {
        return new Location(ToString());
    }

    public override string ToString()
    {
        var items = new char?[] { Side, Vertical, Transverse, Longitudinal };

        var str = new StringBuilder();

        if (Number is not null)
            str.Append(Number.ToString());

        foreach (var item in items)
        {
            if (item is null)
                continue;

            str.Append(item);
        }

        var parts = str.ToString().OrderBy(c => c).Select(s => s.ToString());

        return string.Concat(parts);
    }
}

File: python/vista_sdk/LocationBuilder.py
from __future__ import annotations

from copy import copy as copy
from typing import Optional, TypeVar
from xml.dom import ValidationErr

from vista_sdk.Locations import Location, LocationGroup, Locations

T = TypeVar("T")


class LocationBuilder:
    def __init__(self, locations: Locations):
        self.vis_version = locations.vis_version
        self.reversed_groups = locations._reversed_groups
        self.number = None
        self.side = None
        self.vertical = None
        self.transverse = None
        self.longitudinal = None

    @staticmethod
    def create(locations: Locations) -> LocationBuilder:
        return LocationBuilder(locations)

    def with_location(self, value: Location) -> LocationBuilder:
        builder = copy(self)
        span = str(value.__str__())
        n: Optional[int] = None

        for i, ch in enumerate(span):
            if ch.isdigit():
                if n is None:
                    n = int(ch)
                else:
                    num = Locations.try_parse_int(span, 0, i + 1)
                    if num[0] is False:
                        raise ValidationErr("Should include a valid number")
                    n = num[1]
                continue

            builder = builder.with_value_char(ch)

        if n is not None:
            builder = builder.with_number(n)

        return builder

    def with_number(self, number: int) -> LocationBuilder:
        return self.with_value_internal(LocationGroup.NUMBER, number)

    def with_side(self, side: str) -> LocationBuilder:
        return self.with_value_internal(LocationGroup.SIDE, side)

    def with_vertical(self, vertical: str) -> LocationBuilder:
        return self.with_value_internal(LocationGroup.VERTICAL, vertical)

    def with_transverse(self, transverse: str) -> LocationBuilder:
        return self.with_value_internal(LocationGroup.TRANSVERSE, transverse)

    def with_longitudinal(self, longitudinal: str) -> LocationBuilder:
        return self.with_value_internal(LocationGroup.LONGITUDINAL, longitudinal)

    def with_value(self, value: int) -> LocationBuilder:
        return self.with_value_internal(LocationGroup.NUMBER, value)

    def with_value_char(self, value: str) -> LocationBuilder:
        if value not in self.reversed_groups:
            raise ValueError(f"The value {value} is an invalid Locations value")
        group = self.reversed_groups[value]
        return self.with_value_internal(group, value)

    def with_value_internal(self, group: LocationGroup, value: T) -> LocationBuilder:
        if group == LocationGroup.NUMBER:
            if not isinstance(value, int):
                raise ValueError("Value should be a number")
            if value < 1:
                raise ValueError("Number must be greater than 0")
            builder = copy(self)
            builder.number = value
            return builder

        if not isinstance(value, str) or len(value) != 1:
            raise ValueError("Value should be a single character")

        if value not in self.reversed_groups or self.reversed_groups[value] != group:
            raise ValueError(f"The value {value} is an invalid {group.name} value")

        builder = copy(self)
        setattr(builder, group.name.lower(), value)
        return builder

    def without_value(self, group: LocationGroup) -> LocationBuilder:
        builder = copy(self)
        setattr(builder, group.name.lower(), None)
        return builder

    def build(self) -> Location:
        return Location(self.__str__())

    def __str__(self) -> str:
        parts = [self.side, self.vertical, self.transverse, self.longitudinal]
        parts = [str(p) for p in parts if p is not None]
        if self.number:
            parts.insert(0, str(self.number))
        return "".join(sorted(parts))

    def without_number(self) -> LocationBuilder:
        return self.without_value(LocationGroup.NUMBER)

    def without_side(self) -> LocationBuilder:
        return self.without_value(LocationGroup.SIDE)

    def without_vertical(self) -> LocationBuilder:
        return self.without_value(LocationGroup.VERTICAL)

    def without_transverse(self) -> LocationBuilder:
        return self.without_value(LocationGroup.TRANSVERSE)

    def without_longitudinal(self) -> LocationBuilder:
        return self.without_value(LocationGroup.LONGITUDINAL)
